apps/backend/src/services/cache_service.py

```
@@ -0,0 +1,503 @@
+
+            await self.db.commit()
+            return True
+
+        except Exception as e:
+            await self.db.rollback()
+            logger.error(f"Error updating cache access for {cache_key}: {e}")
+            raise DatabaseOperationError(f"Failed to update cache access: {str(e)}", "update_cache_access")
+
+    async def get_cache_metrics(
+        self,
+        device_ids: Optional[List[UUID]] = None,
```

Ruff (0.12.2)

Use `X | None` for type annotations

Convert to `X | None`

(UP045)